### ISSUE TITLE
Support `/home` mounted with `noexec`

### ISSUE THREAD

**Issue by pillarsdotnet** [pre-commit/pre-commit#1105]
As written, pre-commit assumes that both the git repo and the user's home directory are mounted with `exec` privileges. 

It should be possible to specify an alternate directory where executables may be copied and run.  Alternatively, executables should be symlinked instead of copied, so that they run from their original (presumably `exec`-enabled) locations.

**Comment by asottile** [comment]
set `XDG_CACHE_DIR` or `ln -s ~/.cache/pre-commit wherever` or `PRE_COMMIT_HOME`

**Comment by asottile** [comment]
errr, `XDG_CACHE_HOME` that one -- I always forget the name of that variable

**Comment by pillarsdotnet** [comment]
Thanks; setting `PRE_COMMIT_HOME` and `git --config core.hooksPath` did the trick.  Just gotta automate it next.

**Comment by asottile** [comment]
I've never heard of a person setting noexec on their home directory, I imagine using almost all development tools is painful for you -- glad there's at least a way for you to make this work though :+1:

**Comment by pillarsdotnet** [comment]
Seems to be required for strict auditing standards.

**Comment by asottile** [comment]
seems silly though, for any noexec script you can always just run it with `exectable script.sh` instead (in fact, for the most part `pre-commit` *does* do this, I guess it's just tripping up on the `python` copied executable in virtualenvs and the `.git/hooks/pre-commit` script)

tl;dr I don't think setting `noexec` on home actually increases your security posture :P

**Comment by pillarsdotnet** [comment]
It's not about security; it's about conformance to a security *standard*.  Anyone who has ever gotten certification knows that. ;-)

**Comment by asottile** [comment]
what standard?

**Comment by pillarsdotnet** [comment]
FIPS, DISA, STIG, FEDRAMP, all the government/banking ones.

**Comment by pillarsdotnet** [comment]
(looking...)

I guess the ones I should be concerned with are:
* NIST 800-53
* SOX
* PCI-DSS
* GLBA

Not sure which one has the `/home` requirement.  I'm working with an automated security platform which does most of the work for me.

**Comment by asottile** [comment]
FIPS doesn't mention the setting, couldn't find any information DISA or STIG, FEDRAMP doesn't seem to apply (it's about cloud services) :man_shrugging:

**Comment by pillarsdotnet** [comment]
Well, the server I'm logged into is arguably a "cloud service".  Not gonna try to get pre-commit running on a Windows laptop.

**Comment by asottile** [comment]
it works well on windows :)

**Comment by pillarsdotnet** [comment]
Not when I've got to submit a helpdesk ticket for every single package install, it doesn't.

**Comment by asottile** [comment]
tru
